woocommerce/google-listings-and-ads
# Report endpoints for free listings and campaigns

### Changes proposed in this Pull Request:

This PR adds two reports endpoints for the merchant account (free listings) and ad account (campaigns). In addition to the reports it also refactors the way search queries are run for both Ads and Merchant. The interface is similar to how the DB queries are constructed, but instead it builds GAQL queries and sends a request to the API to retrieve the results.

Closes #236 
Closes #237

### Notes
1. The date selector (copied from WooCommerce Admin) compares the current period with a previous period. That works fine here as well, but it means we need to request the report data for two date ranges (which can be significantly slower since we are requesting data from an API).
2. Because of potential delays in getting report data, should we implement a caching system? Or do we need a pre-fetch method?

### Pending points (still waiting on answers in Q&A doc)
1. The reports from ads allow segmenting by day/week/month/quarter/year whereas the merchant reports only allow segmenting by day
2. I'm not able to retrieve the product name for merchant reports
3. When segmenting by date and by product the merchant reports don't return all products

### Request parameters
Both report endpoints `/gla/mc/reports` or `/gla/ads/reports`support common parameters for retrieving the report data.

#### Report type
The type of report can be specified in the endpoint:
- `reports/programs` - Will segment the data by campaign (for Merchant reports they are all combined under `free_listings`)
- `reports/products` - Will segment the data by product

#### Pagination
This is not a traditional pagination, instead it limits the amount of rows to return and provides a next page token for retrieving the next batch of results. This can be used to breakdown the report data into batches. WCAdmin breaks this down into batches of a 100, I've set the default to 200 (Google will take longer to return the first page, but subsequent pages will be retrieved much faster).
`next_page` will be returned in the response data containing a token to retrieve another page. If this value is null, there are no more pages available.
- `per_page` - Set the limit of rows per page
- `next_page` - Set the next page token for retrieving another page of results

#### Filtering
Filtering can be done by providing a list of ids. Note that we use strings for product ids (currently in the format `gla_<wc_product_id>`). Note that for campaigns we can use either a string or an number, but campaign ids are always a positive integer.
- `ids` - Comma separated list of product or campaign ids

#### Ordering
- `order` - `asc` or `desc`
- `orderby` - Column we would like to order the results by, this must be one of the fields we are requesting

#### Interval
This allows us to specify how we want to segment the data, which is intended for populating a graph. Note that the Merchant data only supports segmenting by day, but the ads report allows the interval to be set to any of the following values:
- `day` - `<year>-<month>-<day>`
- `week` - `<year>-<weeknumber>`
- `month` - `<year>-<month>`
- `quarter` - `<year>-<quarter>`
- `year` - `<year>`

#### Date range
Set the date range we want to retrieve results for, a date can be provided in the format `<year>-<month>-<day>`
- `after` - Defaults to one week ago
- `before` - Defaults to today

### Detailed test instructions:

1. Setup the plugin with a connected merchant account and linked ads account
2. Send a GET request to https://domain.test/wp-json/wc/gla/mc/reports/programs
3. Send a GET request to https://domain.test/wp-json/wc/gla/ads/reports/programs
4. Observe we get valid report information (need a account which has recorded data and has active campaigns)
5. Test additional parameters to the endpoints, such as different report types, pagination, custom order, interval or filtering by id

Note: I have a patch available for a local WCS which retrieves data from accounts with live data. Since this allows retrieving report data through manager requests I didn't include this in a PR.

### Example request:
`GET https://domain.test/wp-json/wc/gla/ads/reports/programs?after=2021-03-01&before=2021-03-14&interval=week&fields[0]=sales&fields[1]=spend&fields[2]=clicks&fields[3]=impressions`
```
{
	"campaigns": [
		{
			"id": 123,
			"name": "One",
			"status": "enabled",
			"subtotals": {
				"sales": 11,
				"spend": 63.14,
				"clicks": 139,
				"impressions": 16037
			}
		},
		{
			"id": 234,
			"name": "Two",
			"status": "enabled",
			"subtotals": {
				"sales": 0,
				"spend": 9.87,
				"clicks": 15,
				"impressions": 901
			}
		}
	],
	"intervals": [
		{
			"interval": "2021-09",
			"subtotals": {
				"sales": 0,
				"spend": 38,
				"clicks": 90,
				"impressions": 9941
			}
		},
		{
			"interval": "2021-10",
			"subtotals": {
				"sales": 11,
				"spend": 35.01,
				"clicks": 64,
				"impressions": 6997
			}
		}
	],
	"totals": {
		"sales": 11,
		"spend": 73.01,
		"clicks": 154,
		"impressions": 16938
	}
}
```

### Changelog Note:
- Report endpoints for free listings and campaigns
